File: day2_part2_2023.cpp
#include <iostream>
#include <fstream>
#include <regex>
#include <sstream>
#include <string>

int main() {
    std::ifstream file("input.txt");
    if (!file.is_open()) {
        std::cout << "Error opening file" << std::endl;
        return 1;
    }

    std::regex gameRegex("Game (\\d+): (.+)");
    std::regex cubeRegex("(\\d+) (red|green|blue)");
    int totalPower = 0;

    std::string line;
    while (std::getline(file, line)) {
        std::smatch matches;
        if (std::regex_search(line, matches, gameRegex)) {
            std::string rounds = matches[2];
            int maxRed = 0, maxGreen = 0, maxBlue = 0;

            std::stringstream ss(rounds);
            std::string round;
            while (std::getline(ss, round, ';')) {
                std::sregex_iterator it(round.begin(), round.end(), cubeRegex);
                std::sregex_iterator end;

                int red = 0, green = 0, blue = 0;
                while (it != end) {
                    int count = std::stoi((*it)[1]);
                    std::string color = (*it)[2].str();

                    if (color == "red") {
                        red += count;
                    } else if (color == "green") {
                        green += count;
                    } else if (color == "blue") {
                        blue += count;
                    }

                    ++it;
                }

                if (red > maxRed) {
                    maxRed = red;
                }
                if (green > maxGreen) {
                    maxGreen = green;
                }
                if (blue > maxBlue) {
                    maxBlue = blue;
                }
            }

            int power = maxRed * maxGreen * maxBlue;
            totalPower += power;
        }
    }

    file.close();

    std::cout << totalPower << std::endl;

    return 0;
}

File: objectivec/day2_part2_2023.m
#import <Foundation/Foundation.h>

int main() {
    NSError *error;
    NSString *fileContents = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:&error];
    NSArray *lines = [fileContents componentsSeparatedByString:@"\n"];
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"Game (\\d+): (.+)" options:0 error:nil];
    NSRegularExpression *cubeRegex = [NSRegularExpression regularExpressionWithPattern:@"(\\d+) (red|green|blue)" options:0 error:nil];
    int totalPower = 0;

    for (NSString *line in lines) {
        NSTextCheckingResult *matches = [regex firstMatchInString:line options:0 range:NSMakeRange(0, [line length])];

        if (matches.numberOfRanges == 3) {
            NSString *rounds = [[line componentsSeparatedByString:@": "] objectAtIndex:1];
            int maxRed = 0, maxGreen = 0, maxBlue = 0;

            NSArray *roundsArray = [rounds componentsSeparatedByString:@";"];

            for (NSString *round in roundsArray) {
                NSArray *cubes = [cubeRegex matchesInString:round options:0 range:NSMakeRange(0, [round length])];
                int red = 0, green = 0, blue = 0;

                for (NSTextCheckingResult *cube in cubes) {
                    int count = [[round substringWithRange:[cube rangeAtIndex:1]] intValue];
                    NSString *color = [round substringWithRange:[cube rangeAtIndex:2]];

                    if ([color isEqualToString:@"red"]) {
                        red += count;
                    } else if ([color isEqualToString:@"green"]) {
                        green += count;
                    } else if ([color isEqualToString:@"blue"]) {
                        blue += count;
                    }
                }

                if (red > maxRed) {
                    maxRed = red;
                }
                if (green > maxGreen) {
                    maxGreen = green;
                }
                if (blue > maxBlue) {
                    maxBlue = blue;
                }
            }

            int power = maxRed * maxGreen * maxBlue;
            totalPower += power;
        }
    }

    printf("%d\n", totalPower);

    return 0;
}